src/services/ftp/backend.rs

```diff
@@ -0,0 +1,512 @@
+
+        if resp.is_file() {
+            meta.set_mode(ObjectMode::FILE);
+        } else if resp.is_directory() {
+            meta.set_mode(ObjectMode::DIR);
+        } else {
+            meta.set_mode(ObjectMode::Unknown);
+        }
+
+        meta.set_content_length(resp.size() as u64);
+
+        meta.set_last_modified(OffsetDateTime::from(resp.modified()));
+
+        Ok(meta)
+    }
+
+    async fn delete(&self, args: &OpDelete) -> Result<()> {
+        let path = args.path();
+        if args.path().ends_with('/') {
+            self.ftp_rmdir(path).await?;
+        } else {
+            self.ftp_rm(path).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn list(&self, args: &OpList) -> Result<DirStreamer> {
+        let path = args.path();
+
+        let rd = self.ftp_list(path).await?;
+
+        Ok(Box::new(DirStream::new(Arc::new(self.clone()), path, rd)))
+    }
+}
+
+impl Backend {
+    pub(crate) fn ftp_connect(&self) -> Result<FtpStream> {
+        // connecting to remote address
+        let u = format! {"{}:{}", self.endpoint, self.port};
+
+        let mut ftp_stream = FtpStream::connect(&u)
+            .map_err(|e| new_request_connection_error("connection", &u, e))?;
+
+        // switch to secure mode if ssl/tls is on.
+        if self.tls {
+            ftp_stream = ftp_stream
+                .into_secure(TlsConnector::new().unwrap(), &u)
+                .map_err(|e| new_request_secure_error("connection", &u, e))?;
+        }
+
+        // login if needed
+        if !self.credential.0.is_empty() {
+            ftp_stream
+                .login(&self.credential.0, &self.credential.1)
+                .map_err(|e| new_request_sign_error("connection", &self.endpoint, e))?;
+        }
+
+        // change to the root path
+        ftp_stream
+            .cwd(&self.root)
+            .map_err(|e| new_request_root_eror("connection", &self.root, e))?;
+
+        Ok(ftp_stream)
+    }
+
+    pub(crate) async fn ftp_get(&self, path: &str) -> Result<Cursor<Vec<u8>>> {
+        let mut ftp_stream = self.ftp_connect()?;
+
+        let reader = ftp_stream
+            .retr_as_buffer(path)
+            .map_err(|e| new_request_retr_error("get", path, e))?;
+
+        ftp_stream
+            .quit()
+            .map_err(|e| new_request_quit_error("get", path, e))?;
+
+        Ok(reader)
+    }
+
+    pub(crate) async fn ftp_put(&self, path: &str) -> Result<u64> {
+        let mut ftp_stream = self.ftp_connect()?;
+
+        let n = ftp_stream
+            .put_file(&path, &mut "".as_bytes())
+            .map_err(|e| new_request_put_error("put", path, e))?;
+        Ok(n)
+    }
+
+    pub(crate) async fn ftp_mkdir(&self, path: &str) -> Result<()> {
+        let mut ftp_stream = self.ftp_connect()?;
+        ftp_stream
+            .mkdir(&path)
+            .map_err(|e| new_request_mkdir_error("mkdir", path, e))?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn ftp_rm(&self, path: &str) -> Result<()> {
```

Comment:
I prefer to write that logic directly. We don't need to extract a new function that only is used once.